pandas/core/util/_dispatching.py

```diff
@@ -0,0 +1,186 @@
+import collections
+import inspect
+from functools import partial
+
+import numpy as np
+
+from pandas._libs import groupby as libgroupby
+from pandas.core.dtypes.missing import isnull
+
+
+class CythonDispatcher(object):
+
+    def __init__(self, groupby):
+        self.groupby = groupby
+        self.func_nm = None
+        self.obj = None
+
+    @property
+    def func_metadata(self):
+        """
+        Stores the metadata required to dispatch each function.
+
+        The format of the dict is as follows:
+
+        attr_name : {
+            'application': {'aggregate', 'transform'}
+            'cython_nm': ...  # Name of the Cython function to call
+            'extra_kwargs': {...}  # Extra kwargs to pass to Cython
+            'type_blacklist': [...]  # Dtypes for which func should raise
+            'result_type': ...  # dtype of result from Cython
+            'conversion_in': ...  # dtype or callable for conversion pre-Cython
+            'needs_values': ...  # Whether the obj values should pass to Cython
+            'needs_mask': ...   # Whether a mask of NA values should be passed
+            'conversion_out': ...  # dtype or callable for conv post-Cython
+        }
+        """
+        return {
+            'any': {
```

Comment:
Thought here is that a dict of dicts can most cleanly describe the metadata that each function may have to pass to Cython